README
ExpectMSU is a Perl module for use internally at Montclair State
University.  It requires the use of Expect.pm to connect to MSU VAX/VMS
systems and MSU Cisco routers.

If you do not have Expect installed, you can use the following command
to download and install it:
	perl -MCPAN -e'install Bundle::Expect'

Installation on a system with a properly configured Perl package should be as
easy as entering "make all" from the top level directory of the distribution.

If you have problems or would like to try the Perl standard method of
installation, cd to the Source directory and follow the same steps as for
the installtion of most other Perl modules:
        1) Use Perl to create the Makefile:
                perl Makefile.pl
        2) Prepare the module:
                make
        3) Install the module:
                make install
        4) Go have a beer.  :)


Changes
Revision history for Perl extension ExpectMSU.

0.20    Tue Oct 28 22:48:57 EDT 2014

    - A couple of minor tweaks (mostly small bugs)
    - Modified some password prompts to account for capitalization

0.12	Mon Aug  6 16:51:12 EDT 2012

	- Added "maximum_buffer_length" setting to adjust the Perl
	  Expect $object->max_accum value

0.111	Thu May  3 12:08:10 EDT 2012

	- Corrected minor error with case of the "Password" prompt when
	  using the UnixSSH connection type
	- Created a "loopAttempts" for how many Expect attempts should
	  be tried

0.11	Mon Feb 20 18:17:34 EST 2012

	- Incorporated manual changes after years(!) of neglect:
		- Added the unixssh connection type
	- Added section for HP switch OS

0.10	Tue Jun  6 14:47:58 EDT 2000
	- first release
	- packaged module

0.01	Tue Jun  6 14:38:44 2000
	- original version; created by h2xs 1.19



Source/test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..1\n"; }
END {print "not ok 1\n" unless $loaded;}
use ExpectMSU;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):



Examples/RunCiscoCommand.pl
#!/bin/sh -- # -*- perl -*-
eval 'exec perl $0 ${1+"$@"}'
      if $running_under_some_shell;

# The above should automatically find Perl for you.  If not, check the first
# line to be something like the following:
#!/usr/local/bin/perl

#### Define debugging if we want it.
$DEBUG=7;

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### Initialization
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

use ExpectMSU;

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### Subroutines
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

sub RunCiscoCmd($$$$) {
  print qq{Entering RunCiscoCmd()\n} if defined $DEBUG;
  (my $cmd, my $ip, my $pwd, my $prompt) = @_;
  print qq{Running: $cmd\n\ton: IP=$ip; Pwd=$pwd; Prompt=$prompt\n}
    if defined $DEBUG;
  my $router = ExpectMSU->new(
    debug    => ($DEBUG - 1),
    hostname => "$ip",
    ostype   => "Cisco",
    password => "$pwd",
    prompt   => "$prompt",
    pager    => "--More--"
  );
  return $router->execute($cmd);
}

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### Main
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

# Get information
$ip = shift or die qq{Specify the IP address to connect to.\n};
$pass = shift or die qq{Specify the password to authenticate with.\n};
$prompt = shift or die qq{Specify the prompt the router uses.\n};
$command = shift or die qq{Specify the command to run.\n};

# Split the output into an arrary, one line per element
print RunCiscoCmd($command, $ip, $pass, $prompt);
# @table = split(/[\n\r]+/, RunCiscoCmd($command, $ip, $pass, $prompt));

# Print eveything out
print "\n" . join("\n", @table) . "\n";


Examples/GetCiscoArpTable.pl
#!/bin/sh -- # -*- perl -*-
eval 'exec perl $0 ${1+"$@"}'
      if $running_under_some_shell;

# The above should automatically find Perl for you.  If not, check the first
# line to be something like the following:
#!/usr/local/bin/perl

#### Define debugging if we want it.
#$DEBUG=3;

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### User settings
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

# Define the settings for the routers we will collect arp data from
@ROUTERS = (
  "putRouterIpHere|putPassHere|putRouterPromptHere",
);

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### Initialization
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

use ExpectMSU;

# Get the current timestamp
#  0    1    2     3     4    5     6     7     8
#($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst)
my @time = localtime();
$time[4] += 1;
$time[5] += 1900;
my $timestamp = sprintf("%d/%d/%d %d:%d:00", $time[4], $time[3], $time[5],
                                             $time[2], $time[1]);

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### Subroutines
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

sub GetCurrentInfo(@) {
  print qq{Entering GetCurrentInfo()\n} if defined $DEBUG;
  my @routers = @_;
  print qq{Received:\n\t} . join ("\n\t", @routers) . "\n" if defined $DEBUG;
  my %ip2mac;
  foreach $router (@routers) {
    my %temp = GetRouterTable($router);
    foreach my $ip (keys %temp) {
      if (not defined $ip2mac{$ip}) {
        $ip2mac{$ip} = $temp{$ip};
      } else {
        die qq{IP ($ip) defined twice ($ip2mac{$ip} and $temp{$ip})}
          unless ($ip2mac{$ip} eq $temp{$ip});
      }
    }
  }
  return %ip2mac;
}

sub GetRouterTable($) {
  print qq{Entering GetRouterTable()\n} if defined $DEBUG;
  my %ip2mac;
  my $cmd = qq{show ip arp};
  my @table = split(/[\n|\r]/, RunCiscoCmd($cmd, split(/\|/, shift)));
  print qq{Found:\n\t}.join("\n\t",@table).qq{\n} if ($DEBUG > 2);
  # Parse the arp table
  foreach (@table) {
    chomp;
    (my $proto, my $ip, my $age, my $mac, my $type, my $int) = split;
    next unless $ip =~ /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/;
    next unless $mac =~ /[\da-f]{4}\.[\da-f]{4}\.[\da-f]{4}/i;
    $ip = sortableIP($ip);
    next if ($mac eq "0000.0000.0000");
    die qq{Invalid mac ($mac)} if (length($mac) != 14);
    $mac = prettyMAC($mac);
    if (not defined $ip2mac{$ip}) {
      $ip2mac{$ip} = $mac;
    } else {
      die qq{IP ($ip) defined twice ($ip2mac{$ip} and $mac)}
        unless ($ip2mac{$ip} == $mac);
    }
  }
  return %ip2mac;
}

sub RunCiscoCmd($$$$) {
  print qq{Entering RunCiscoCmd()\n} if defined $DEBUG;
  (my $cmd, my $ip, my $pwd, my $prompt) = @_;
  print qq{Running: $cmd\n\ton: IP=$ip; Pwd=$pwd; Prompt=$prompt\n}
    if defined $DEBUG;
  my $router = ExpectMSU->new(
    debug    => ($DEBUG - 1),
    hostname => "$ip",
    ostype   => "Cisco",
    password => "$pwd",
    prompt   => "$prompt",
    pager    => "--More--"
  );
  return $router->execute($cmd);
}

sub cleanMAC($) {
  my $MAC = uc shift;
  $MAC =~ tr/[0-9][A-F]//cd;
  return $MAC;
}

sub prettyMAC($) {
  return sprintf("%02s:%02s:%02s:%02s:%02s:%02s",
           unpack("A2A2A2A2A2A2",
             cleanMAC(shift)
           )
         );
}

sub cleanIP($) {
  my $IP = shift;
  $IP =~ tr/[0-9]\.//cd;
  return $IP
}

sub sortableIP($) {
  return sprintf("%3d.%3d.%3d.%3d", split(/\./, cleanIP(shift)));
}

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### Main
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

print qq{Loading current arp table.\n} if defined $DEBUG;
%Arp = GetCurrentInfo(@ROUTERS);

print qq{Comparing current entries to previous entries.\n} if defined $DEBUG;
foreach $ip (sort keys %Arp) {
  print cleanIP($ip) . "\t" . prettyMAC($Arp{$ip}) . "\n";
}


Legacy_Versions/ExpectVMS/original.pl
#!/usr/bin/perl

#$DEBUG=1;

$Home="/home/data/VMS_Data";

# Alpha DNS hostname or IP address
$AlphaAddress="Alpha.Montclair.edu";
# Username and password to telnet/FTP to Alpha with.
$AlphaUser="data"; $AlphaPass="itdata00";
# Remote and local names to use for the SIS text file
$AlphaRemoteSIS = "SIS"; $AlphaSIS = "$Home/SIS.txt";
# Remote and local names to use for the UAF text file
$AlphaRemoteUAF = "SYSUAF.LIS"; $AlphaUAF = "$Home/AlphaUAF.txt";

# Saturn DNS hostname or IP address
$SaturnAddress="Saturn.Montclair.edu";
# Username and password to telnet/FTP to Saturn with.
$SaturnUser="data"; $SaturnPass="itdata00";
# Remote and local names to use for the UAF text file
$SaturnRemoteUAF = "SYSUAF.LIS"; $SaturnUAF = "$Home/SaturnUAF.txt";

use Expect;
use IO::File;
use Net::FTP;

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### Main script
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

# Run some commands on Alpha
ExecVMS ($AlphaAddress, $AlphaUser, $AlphaPass,
         # Purge the directory.
         "PURGE /KEEP=1 *.*",
         # Do a directory listing
         "DIRECTORY /DATE /SIZE",
         # Delete log files
         "DELETE *.LOG;*",
         # Create the UAF listing.
         "MC AUTHORIZE LIST * /FULL");

# Run some commands on Saturn
ExecVMS ($SaturnAddress, $SaturnUser, $SaturnPass,
         # Purge the directory.
         "PURGE /KEEP=1 *.*",
         # Do a directory listing
         "DIRECTORY /DATE /SIZE",
         # Delete log files
         "DELETE *.LOG;*",
         # Create the UAF listing.
         "MC AUTHORIZE LIST * /FULL");

# FTP to Alpha
$ftp = Net::FTP->new($AlphaAddress, Debug => $DEBUG+0);
$ftp->login("$AlphaUser","$AlphaPass");

# Get the SIS file
$SIS = new IO::File "> $AlphaSIS";
if (defined $SIS) {
  $ftp->get($AlphaRemoteSIS, $SIS);
  $SIS->close;
}

# Get the UAF file
$UAF = new IO::File "> $AlphaUAF";
if (defined $UAF) {
  $ftp->get($AlphaRemoteUAF, $UAF);
  $UAF->close;
}

# Close the FTP connection
$ftp->quit;

# FTP to Saturn
$ftp = Net::FTP->new($SaturnAddress, Debug => $DEBUG+0);
$ftp->login("$SaturnUser","$SaturnPass");

# Get the UAF file
$UAF = new IO::File "> $SaturnUAF";
if (defined $UAF) {
  $ftp->get($SaturnRemoteUAF, $UAF);
  $UAF->close;
}

# Close the FTP connection
$ftp->quit;

print "\n\n***** Done! *****\n\n" if defined $DEBUG;
exit;

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
####
#### Subroutine to run commands on a remote VMS system via telnet
####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

sub ExecVMS() {

  # Define variables
  my $prompt = '\$\s+$';
  (my $host, my $user, my $pass)= splice (@_, 0, 3);

  @cmds = @_;

  # Open the telnet session
  $telnet = Expect->spawn("telnet $host") or die "Can't telnet $host";
  $telnet->log_stdout(0) unless defined $DEBUG;

  # Time out if we don't get username within 10 seconds.
  unless ($telnet->expect(10,"Username: ")) {
    die "Never got username prompt" . $telnet->exp_error()."\n";
  }
  print $telnet "$user\r";

  # Time out if we don't get password within 10 seconds.
  unless ($telnet->expect(10,"Password: ")) {
    die "Never got password prompt" . $telnet->exp_error()."\n";
  }
  print $telnet "$pass\r";

  # First, make sure the prompt is set to something we recognize
  while (not($telnet->expect(15, "-re", "$prompt"))) {
    die "Never found prompt" if ($loop++ gt 8);
    print $telnet "SET PROMPT\r";
  }

  # Now that we're in and have a known prompt, let's get to work.
  foreach $cmd (@cmds) {
    print STDERR qq{\n***** Running: "$cmd"\n} if defined $DEBUG;
    print $telnet $cmd . "\r";
    die "Never got prompt back" unless $telnet->expect(600, "-re", "$prompt");
  }

  # Close up shop and go home
  $cmd = "LOGOUT /FULL";
  print STDERR qq{\n***** Running: "$cmd"\n} if defined $DEBUG;
  print $telnet $cmd . "\r";
  $telnet->soft_close();
  return 0;
}


Source/ExpectMSU.pm
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

package ExpectMSU;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;
require AutoLoader;

# Require the Expect.pm package
use Expect; # or die qq{Expect.pm MUST be installed!};;

@ISA = qw(Exporter AutoLoader);
@EXPORT = qw();
@EXPORT_OK = qw();
$VERSION = '0.20';

my %settings = (
  debug => 0,
  hostname => "localhost",
  ostype => "",
  username => "",
  password => "",
  timeout => 15,
  loopAttempts => 2,
  prompt => "j u n k",
  pager => "j u n k",
  expect => "",
  maximum_buffer_length => "655360"
);

# Preloaded methods go here.
# Autoload methods go after =cut, and are processed by the autosplit program.

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
####
#### Routines to create new connections.
####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
sub new {

  # Determine our class.
  my $class = shift;

  # Create a new object.
  print STDERR qq{Creating a new object.\n} if ($settings{debug} > 0);
  my $self = bless {}, $class;

  # Parse any settings.
  print STDERR qq{Reading settings.\n} if ($settings{debug} > 0);
  $self->settings (@_);

  # Initialize the connection.
  print STDERR qq{Initializing connection.\n} if ($settings{debug} > 0);
  $self->_initialize;

  # We're done.
  print STDERR qq{Object creation complete.\n} if ($settings{debug} > 0);
  return $self;
}

sub settings {

  my $self = shift;

  # Get function parameters.
  print STDERR qq{Reading parameters.\n} if ($settings{debug} > 0);
  my %hash = @_;
  my $key;
  foreach $key (keys %hash) {
    if (defined $settings{$key}) {
      $settings{$key} = $hash{$key};
    } else {
      die qq{Invalid option "$key"};
    }
  }

  # Check the timeout value.
  die ("Invalid value for timeout") unless ($settings{timeout} > 0);

  # Check the ostype.
  $settings{ostype} = lc $settings{ostype};

  # Check the debug value (make sure it's numeric and under 3).
  $settings{debug} += 0;
  $settings{debug}++ while ($settings{debug} < 0);
  $settings{debug}-- while ($settings{debug} > 3);

  # Check the timeout value (make sure it's numeric and at least 5).
  $settings{timeout} += 0;
  $settings{timeout}++ while ($settings{timeout} < 4);

  if ($settings{debug} > 0) {
    print STDERR "Settings:\n";
    foreach $key (keys %settings) {
      print STDERR "$key: $settings{$key}\n";
    }
    print STDERR "\n";
  }

  return 0;
}

sub execute {
  my $self = shift;
  my $cmd = shift;

  $self->settings (@_) if (@_);

  my $expect = $settings{expect};
  #$expect->slave->stty(qw(raw));

  print STDERR qq{\n***** Running: "$cmd"\n} if ($settings{debug} > 0);
  print $expect "$cmd\r";
  my $match; my $loop = 1; my $count; my $output; my @output; my @text;
  my $output = "";
  while ($loop) {
    $match = $expect->expect( $settings{timeout},
                                "-re", "$settings{prompt}",
                                "-re", "$settings{pager}");
    die "Never got prompt or pager.\n" unless $match;
    $output .= $expect->exp_before();
    _printExpectDebug("prompt", $output);
    $loop = $match - 1;
    $count++;
    print STDERR "Waiting for prompt ($count)\n" if $settings{debug};
    print $expect " " if $loop;
  }
  $output .= $expect->clear_accum();
  my @output = split(/\n/, $output);
  push (@text, @output);
  my $results = join("\n", @text) . "\n";
  $results =~ s/\e\[7m\n\e\[0m\010+//g;
  return $results;
}

sub close {
  my $self = shift;
  $self->DESTROY;
}

sub DESTROY {
  my $self = shift;

  # Close the connection for the given OS.
  print STDERR qq{Determining OS type.\n} if ($settings{debug} > 0);
  if ($settings{ostype} eq "vms") {
    print STDERR qq{VMS OS type.\n} if ($settings{debug} > 0);
    $self->_close_vms;
  } elsif ($settings{ostype} eq "cisco" ) {
    print STDERR qq{Cisco OS via Telnet type.\n} if ($settings{debug} > 0);
    $self->_close_cisco;
  } elsif ($settings{ostype} eq "ciscossh" ) {
    print STDERR qq{Cisco OS via SSH type.\n} if ($settings{debug} > 0);
    $self->_close_ciscossh;
  } elsif ($settings{ostype} eq "unixssh" ) {
    print STDERR qq{Unix SSH type.\n} if ($settings{debug} > 0);
    $self->_close_unixssh;
  } elsif ($settings{ostype} eq "merussh" ) {
    print STDERR qq{Meru SSH type.\n} if ($settings{debug} > 0);
    $self->_close_unixssh;
  } elsif ($settings{ostype} eq "hpswitchos" ) {
    print STDERR qq{HP switch OS via telnet type.\n} if ($settings{debug} > 0);
    $self->_close_hpswitch;
  } else {
    die qq{Invalid OS type defined "$settings{ostype}"};
  }
}

sub _initialize {

  my $self = shift;

  # Initialize the connection for the given OS.
  print STDERR qq{Determining OS type.\n} if ($settings{debug} > 0);
  if ($settings{ostype} eq "vms") {
    print STDERR qq{VMS OS type.\n} if ($settings{debug} > 0);
    $self->_initialize_vms;
  } elsif ($settings{ostype} eq "cisco" ) {
    print STDERR qq{Cisco OS via Telnet type.\n} if ($settings{debug} > 0);
    $self->_initialize_cisco;
    print STDERR qq{Running "term len 0"\n} if ($settings{debug} > 0);
    $self->execute("term len 0");
  } elsif ($settings{ostype} eq "ciscossh" ) {
    print STDERR qq{Cisco OS via SSH type.\n} if ($settings{debug} > 0);
    $self->_initialize_ciscossh;
    print STDERR qq{Running "term len 0"\n} if ($settings{debug} > 0);
    $self->execute("term len 0");
  } elsif ($settings{ostype} eq "unixssh" ) {
    print STDERR qq{Unix via SSH type.\n} if ($settings{debug} > 0);
    $self->_initialize_unixssh;
  } elsif ($settings{ostype} eq "merussh" ) {
    print STDERR qq{Meru via SSH type.\n} if ($settings{debug} > 0);
    $self->_initialize_merussh;
  } elsif ($settings{ostype} eq "hpswitchos" ) {
    print STDERR qq{HP switch OS via telnet type.\n} if ($settings{debug} > 0);
    $self->_initialize_hpswitch;
  } else {
    die qq{Invalid OS type defined "$settings{ostype}"};
  }

  # Define settings for the connection.
  my $expect = $settings{expect};
  my $maximum_buffer_length = $settings{maximum_buffer_length};
  if ($maximum_buffer_length > 0) {
    my $current_buffer_length = $expect->max_accum();
    print STDERR qq{Setting max_accum (maximum_buffer_length) to "$maximum_buffer_length" from "$current_buffer_length".\n} if ($settings{debug} > 0);
    $expect->max_accum($maximum_buffer_length);
  }

}

sub _initialize_hpswitch {

  my $self = shift;
  my $expect;

  # Open the telnet session
  print STDERR "Running telnet.\n" if $settings{debug};
  $expect = Expect->spawn("telnet $settings{hostname}")
    or die "Can't telnet $settings{hostname}\n";
  $settings{expect} = $expect;
  $expect->log_stdout(0);

  print STDERR "Waiting for password prompt.\n" if $settings{debug};
  unless ( $expect->expect( $settings{timeout},"Enter password: ")) {
    die "Never got password prompt " . $expect->exp_error()."\n";
  }
  print $expect "$settings{password}\r";
  _printExpectDebug("Enter password: ", $expect->exp_before());

  # First, make sure the prompt is set to something we recognize
  print STDERR "Waiting for prompt.\n" if $settings{debug};
  my $loop=0;
  while (not( $expect->expect( $settings{timeout}, "-re", $settings{prompt}))) {
    die "Never found $settings{prompt}.\n" if ($loop++ > $settings{loopAttempts});
    print $expect "\r";
  }
  _printExpectDebug("$settings{prompt}", $expect->exp_before());
}

sub _close_hpswitch {
  my $self = shift;

  my $cmd = "exit";
  my $expect=$settings{expect};

  print STDERR qq{\n***** Running: "$cmd"\n} if $settings{debug};
  print $expect "$cmd\r";
  $expect->soft_close();
  $settings{expect} = undef;
  return 0;
}





sub _initialize_cisco {

  my $self = shift;
  my $expect;

  # Open the telnet session
  print STDERR "Running telnet.\n" if $settings{debug};
  $expect = Expect->spawn("telnet $settings{hostname}")
    or die "Can't telnet $settings{hostname}\n";
  $settings{expect} = $expect;
  $expect->log_stdout(0);

  if ($settings{'username'}) {
    print STDERR "Waiting for username prompt.\n" if $settings{debug};
    unless ( $expect->expect( $settings{timeout},"Username: ")) {
      die "Never got username prompt " . $expect->exp_error()."\n";
    }
    print $expect "$settings{username}\r";
    _printExpectDebug("Username: ", $expect->exp_before());
  }

  my $pass_prompt = "assword: ";
  print STDERR "Waiting for password prompt ($pass_prompt).\n" if $settings{debug};
  unless ( $expect->expect( $settings{timeout},$pass_prompt)) {
    die "Never got password prompt ($pass_prompt) " . $expect->exp_error()."\n";
  }
  print $expect "$settings{password}\r";
  _printExpectDebug("Password: ", $expect->exp_before());

  # First, make sure the prompt is set to something we recognize
  print STDERR "Waiting for prompt (" . $settings{prompt} . ").\n" if $settings{debug};
  my $loop=0;
  while (not( $expect->expect( $settings{timeout}, "-re", $settings{prompt}))) {
    die "Never found $settings{prompt}.\n" if ($loop++ > $settings{loopAttempts});
    print $expect "\r";
  }
  _printExpectDebug("$settings{prompt}", $expect->exp_before());
}

sub _close_cisco {
  my $self = shift;

  my $cmd = "exit";
  my $expect=$settings{expect};

  print STDERR qq{\n***** Running: "$cmd"\n} if $settings{debug};
  print $expect "$cmd\r";
  $expect->soft_close();
  $settings{expect} = undef;
  return 0;
}





sub _initialize_ciscossh {

  my $self = shift;
  my $expect;

  # Open the telnet session
  print STDERR "Running SSH.\n" if $settings{debug};
  my $cmd = "ssh $settings{username}\@$settings{hostname}";
  print STDERR qq{Command: $cmd\n} if ($settings{debug} > 0);
  $expect = Expect->spawn($cmd)
    or die "Can't SSH!\n";
  $settings{expect} = $expect;
  $expect->log_stdout(0);

  print STDERR "Waiting for password prompt.\n" if $settings{debug};
  my $pass_prompt = "assword: ";
  unless ( $expect->expect( $settings{timeout},$pass_prompt)) {
    die "Never got password prompt ($pass_prompt) " . $expect->exp_error()."\n";
  }
  print $expect "$settings{password}\r";
  _printExpectDebug("Password: ", $expect->exp_before());

  # First, make sure the prompt is set to something we recognize
  print STDERR "Waiting for prompt.\n" if $settings{debug};
  my $loop=0;
  while (not( $expect->expect( $settings{timeout}, "-re", $settings{prompt}))) {
    die "Never found $settings{prompt}.\n" if ($loop++ > $settings{loopAttempts});
    print $expect "\r";
  }
  _printExpectDebug("$settings{prompt}", $expect->exp_before());
}

sub _close_ciscossh {
  my $self = shift;

  my $cmd = "exit";
  my $expect=$settings{expect};

  print STDERR qq{\n***** Running: "$cmd"\n} if $settings{debug};
  print $expect "$cmd\r";
  $expect->soft_close();
  $settings{expect} = undef;
  return 0;
}





sub _initialize_unixssh {

  my $self = shift;
  my $expect;

  # Open the telnet session
  print STDERR "Running SSH.\n" if $settings{debug};
  $expect = Expect->spawn("ssh $settings{username}\@$settings{hostname}")
    or die "Can't SSH $settings{username}\@$settings{hostname}\n";
  $settings{expect} = $expect;
  $expect->log_stdout(0);

  print STDERR "Waiting for password prompt.\n" if $settings{debug};
  unless ( $expect->expect( $settings{timeout},"Password: ")) {
    die "Never got password prompt " . $expect->exp_error()."\n";
  }
  print $expect "$settings{password}\r";
  _printExpectDebug("Password: ", $expect->exp_before());

  # First, make sure the prompt is set to something we recognize
  print STDERR "Waiting for prompt.\n" if $settings{debug};
  my $loop=0;
  while (not( $expect->expect( $settings{timeout}, "-re", $settings{prompt}))) {
    die "Never found $settings{prompt}.\n" if ($loop++ > $settings{loopAttempts});
    print $expect "\r";
  }
  _printExpectDebug("$settings{prompt}", $expect->exp_before());
}

sub _close_unixssh {
  my $self = shift;

  my $cmd = "exit";
  my $expect=$settings{expect};

  print STDERR qq{\n***** Running: "$cmd"\n} if $settings{debug};
  print $expect "$cmd\r";
  $expect->soft_close();
  $settings{expect} = undef;
  return 0;
}










sub _initialize_merussh {

  my $self = shift;
  my $expect;

  # Open the telnet session
  print STDERR "Running SSH.\n" if $settings{debug};
  $expect = Expect->spawn("ssh $settings{username}\@$settings{hostname}")
    or die "Can't SSH $settings{username}\@$settings{hostname}\n";
  $settings{expect} = $expect;
  $expect->log_stdout(0);

  print STDERR "Waiting for password prompt.\n" if $settings{debug};
  unless ( $expect->expect( $settings{timeout},"password: ")) {
    die "Never got password prompt " . $expect->exp_error()."\n";
  }
  print $expect "$settings{password}\r";
  _printExpectDebug("Password: ", $expect->exp_before());

  # First, make sure the prompt is set to something we recognize
  print STDERR "Waiting for prompt.\n" if $settings{debug};
  my $loop=0;
  while (not( $expect->expect( $settings{timeout}, "-re", $settings{prompt}))) {
    die "Never found $settings{prompt}.\n" if ($loop++ > $settings{loopAttempts});
    print $expect "\r";
  }
  _printExpectDebug("$settings{prompt}", $expect->exp_before());
}

sub _close_merussh {
  my $self = shift;

  my $cmd = "exit";
  my $expect=$settings{expect};

  print STDERR qq{\n***** Running: "$cmd"\n} if $settings{debug};
  print $expect "$cmd\r";
  $expect->soft_close();
  $settings{expect} = undef;
  return 0;
}










sub _initialize_vms {

  my $self = shift;
  my $expect;

  $settings{prompt} = '\$\s+$';

  # Open the telnet session
  print STDERR "Running telnet.\n" if $settings{debug};
  $expect = Expect->spawn("telnet $settings{hostname}")
    or die "Can't telnet $settings{hostname}\n";
  $settings{expect} = $expect;
  $expect->log_stdout(0);

  # Time out if we don't get username.
  print STDERR "Waiting for Username: prompt.\n" if $settings{debug};
  unless ( $expect->expect( $settings{timeout},"Username: ")) {
    die "Never got username prompt" . $expect->exp_error() . "\n";
  }
  print $expect "$settings{username}\r";
  _printExpectDebug("Username: ", $expect->exp_before());

  print STDERR "Waiting for Password: prompt.\n" if $settings{debug};
  unless ( $expect->expect( $settings{timeout},"Password: ")) {
    die "Never got password prompt" . $expect->exp_error()."\n";
  }
  print $expect "$settings{password}\r";
  _printExpectDebug("Password: ", $expect->exp_before());

  print STDERR "Waiting for prompt.\n" if $settings{debug};
  # First, make sure the prompt is set to something we recognize
  my $loop=0;
  my $delay = int($settings{timeout}/10);
  while (not( $expect->expect($delay, "-re", $settings{prompt}))) {
    die "Never found prompt.\n" if ($loop++ > $settings{loopAttempts});
    print STDERR "Loop pass $loop ($delay seconds).\n"
      if ($settings{debug} > 0);
    print $expect "SET PROMPT\r";
  }
  _printExpectDebug("\$", $expect->exp_before());
}

sub _close_vms {
  my $self = shift;

  my $cmd = "LOGOUT /FULL";
  my $expect=$settings{expect};

  print STDERR qq{\n***** Running: "$cmd"\n} if $settings{debug};
  print $expect "$cmd\r";
  $expect->soft_close();
  $settings{expect} = undef;
  return 0;
}

sub _printExpectDebug {

  my $search = shift @_;
  my $result = shift @_;

  if ($settings{debug} > 1) {
    print STDERR qq{**** Received the following while waiting for "$search".\n};
    my $line;
    my $char;
    foreach $line (split(/\n/, $result)) {
      $line =~ s/[\r\n]*$//;
      foreach $char (split(//, $line)) {
        if (not ((32 <= ord($char)) and (ord($char) <= 126))) {
          $char = ord($char);
          $char = "ESC" if $char eq 27;
          $char = "CR" if $char eq 13;
          $char = "<" . $char . ">";
        }
        print STDERR $char;
      }
      print STDERR "\n";
    }
    print STDERR qq{**** Found "$search".\n};
  }
}

1;
__END__

#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
####
#### Below is the stub of documentation for your module. You better edit it!
####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####
#### #### #### #### #### #### #### #### #### #### #### #### #### #### #### ####

=head1 NAME

ExpectMSU - Perl/Expect object to execute commands on a remote system

=head1 SYNOPSIS

  use ExpectMSU;

=head1 DESCRIPTION

This extension provides an easy interface to execute commands
on a remote system via telnet.

$router = ExpectMSU->new(
    debug    => "$DEBUG",
    hostname => "$ip",
    ostype   => "Cisco",
    password => "$pwd",
    prompt   => "$prompt",
    pager    => "--More--"
);

print $router->execute("show ip int br");

=head1 AUTHOR

Brian Kelly, Brian.Kelly@Montclair.edu

=head1 SEE ALSO

perl(1).

=cut
